Fix window counter (exclude hidden windows)

GeoffreyCoulaud opened the pull request:
Hide windows that hint not to be shown in the task bar.
Fix for #1629

slavicd reviewed (approved):
Ran in Fedora 35

kanakshilledar commented:
![nautilus window counter](https://user-images.githubusercontent.com/56501211/149985314-315f5c64-fd02-43e4-814e-bb8468714aeb.png)
Sorry, not working for me. I tried opening two nautilus windows and then tried copying one file to the other then after the transfer process ended the counter incremented by 2. I have attached a screenshot for reference.
tried logging out.

GeoffreyCoulaud commented:
@kanakshilledar I could not reproduce, following your procedure I have only 2 nautilus indicators before, during and after the copy.  

Did you test this MR properly ?   
You need to disable the extension, build and install my branch, reload `gnome-shell` (`alt`+`f2`, `r`, `return`) and then try again. 

If it's still broken on your machine, I'll need to investigate deeper.

slavicd commented:
I only needed the restart gnome shell part for the changes to take effect. I'll restate that the fix worked for me (I am, in fact, still running the code until the upstream merges this).

kanakshilledar commented:
@GeoffreyCoulaud  I did the following
- Removed the old extension from gnome-extension using the `remove` option
- Cloned your branch with `$ git clone https://github.com/GeoffreyCoulaud/dash-to-dock`
- Installed by `$ make` and `$ make install`
- reloaded the shell using `alt` + `f2`, `r`)
But still I am having that issue.
I am having Gnome shell version 41.3 and Endeavour OS (arch linux)
I have also attached a sample video [video_link](https://youtu.be/BVQG-aAdhW0)

kanakshilledar commented:
@GeoffreyCoulaud  I am using X11 and getting this issue.
```
$ loginctl show-session 7 -p Type
Type=x11 
```

GeoffreyCoulaud commented:
Yup, I tested again and It's broken in both now. I'm reading the mutter docs to see what's going wrong.

GeoffreyCoulaud commented:
@kanakshilledar You may want to report that to the EndeavourOS maintainers.  
I've checked with my Manjaro install, mutter is the same version so it's likely not that.  
My fix works on Arch, Manjaro and Fedora.

I've also tried filtering with the `get_window_type()` function and only keeping `NORMAL`, `DIALOG` and `MODAL_DIALOG` ([docs here](https://gjs-docs.gnome.org/meta9~9_api/meta.windowtype)) and even that is not working. Maybe it's somewhere earlier in the call stack that breaks ? I have no idea.

kanakshilledar commented:
Oh okay I will try reporting them... Thanks for the suggestion

kanakshilledar commented:
I tried Reborn OS and in the live image I added the extension and it was working fine only once I felt that issue. So maybe it's with Endeavour OS only

3v1n0 commented on appIcons.js:
```diff
@@ -1233,8 +1233,11 @@ function getInterestingWindows(windows, monitorIndex) {
         windows = windows.filter(function(w) {
-            return w.get_workspace().index() == activeWorkspace
-                || (showUrgent && (w.urgent || w.demandsAttention || w._manualUrgency));
+            // See GJS docs : Meta/Window
```
This is not needed, it should be pretty known for who is into developing a shell extension :)

3v1n0 commented on appIcons.js:
```diff
@@ -1233,8 +1233,11 @@ function getInterestingWindows(windows, monitorIndex) {
+            // See GJS docs : Meta/Window
+            const inWorkspace = w.get_workspace().index() == activeWorkspace;
+            const isUrgent = w.urgent || w.demandsAttention || w._manualUrgency;
+            const isNotInTaskbar = w.is_skip_taskbar();
```
I was thinking to include this for some time now too... I've been always wondering if this would break the case of showing the windows into the previews-menu though.

In fact while the windows may not be to be shown in the lists, *maybe* we may want to show them in the previews overview... But I'm not either sure about :)

3v1n0 reviewed (changes requested):
idea looks ok though, but I think it must be at upper level, so outside the `settings.get_boolean('isolate-workspaces')` check scope.

Ideally you can just `return windows.filter(w => !w.skipTaskbar);`

3v1n0 commented:
Ah, and please remove the duplicated filter on `_rebuildMenu`.

GeoffreyCoulaud commented:
Thanks for the review.  
I implemented the changes, I also added curly braces to the `isolate_monitors` check to be more consistent.  

About the ability to separate visibility in the indicator and in the preview, I'm not sure about that. I think having a different number of indicators and previews in unexpected for the user. Having options is great, still I believe the added complexity and the difference from the overview dock is not worth it.

3v1n0 commented on appIcons.js:
```diff
@@ -1233,17 +1232,19 @@ function getInterestingWindows(windows, monitorIndex) {
         windows = windows.filter(function(w) {
-            return w.get_workspace().index() == activeWorkspace
-                || (showUrgent && (w.urgent || w.demandsAttention || w._manualUrgency));
+            const inWorkspace = w.get_workspace().index() == activeWorkspace;
```
Since you're there, can you please change to `===` equality check?

GeoffreyCoulaud replied:
Will do !

3v1n0 commented on appIcons.js:
```diff
@@ -1233,17 +1232,19 @@ function getInterestingWindows(windows, monitorIndex) {
     }
 
-    if (settings.get_boolean('isolate-monitors'))
+    if (settings.get_boolean('isolate-monitors')){
```
```suggestion
    if (settings.get_boolean('isolate-monitors')) {
```

3v1n0 commented on appIcons.js:
```diff
@@ -1233,17 +1232,19 @@ function getInterestingWindows(windows, monitorIndex) {
+    }
 
-    return windows;
+    return windows.filter(w => !w.is_skip_taskbar());
```
I like using the property better here:

```suggestion
    return windows.filter(w => !w.skipTaskbar);
```

FreddieChopin commented:
Sorry for noob question, but is the thing you discuss above the same as the issue I'm seeing right now:
![Screenshot from 2022-01-31 15-20-45](https://user-images.githubusercontent.com/1139038/151811030-a5875f8a-7aab-4b74-8f4c-20d754276abc.png)
I have just one Firefox window opened (I had two, but I closed it, yet there are 2 dots), only 1 window from e-mail client (sent the e-mail I was writing, so these additional windows are closed too), 1 windows of Nautilus.

kanakshilledar commented:
I had the issue but it got fixed by the recent update.

sergiotca commented:
> Sorry for noob question, but is the thing you discuss above the same as the issue I'm seeing right now: ![Screenshot from 2022-01-31 15-20-45](https://user-images.githubusercontent.com/1139038/151811030-a5875f8a-7aab-4b74-8f4c-20d754276abc.png) I have just one Firefox window opened (I had two, but I closed it, yet there are 2 dots), only 1 window from e-mail client (sent the e-mail I was writing, so these additional windows are closed too), 1 windows of Nautilus.

Yep, this is the issue at hand here.

sergiotca commented:
@FreddieChopin Yep, that is the issue at hand here.

sergiotca commented:
> I had the issue but it got fixed by the recent update.

What update are you talking about buddy?

GeoffreyCoulaud commented:
@sergiotca Probably their distribution's package managers updated the extension.

To anyone unsure if the report applies to their bug : If it's about more indicators than should be, it's this issue and if you compile the extension from source you'll see it being resolved.

tareqimbasher commented:
Thanks guys. When can we expect this fix to be pushed to extensions.gnome.org?

kanakshilledar commented:
@sergiotca I manually compiled the Dash to Dock extension and the issue got resolved as suggested by @GeoffreyCoulaud.
